src/plugins/discover/public/application/main/services/discover_state.ts

```diff
@@ -319,6 +320,10 @@ export function getState({
         stateStorage
       );
 
+      // some filters may not be valid for this context, so update
+      // the filter manager with a modified list of valid filters
+      filterManager.setFilters(getValidFilters(dataView, filterManager.getFilters()));
```

Comment:
One last ask, how about only executing `setFilters` if necessary? 
```suggestion
      const currentFilters = filterManager.getFilters();
      const validFilters = getValidFilters(dataView, currentFilters);
      if (!isEqual(currentFilters, validFilters)) {
        filterManager.setFilters(validFilters);
      }
```